### PR TITLE
Initial support for docker local development

### PR DESCRIPTION
**Pull request by silviot**
## Description
This PR adds a new command `dallinger docker debug` which behaves similarly to `dallinger debug` but uses docker compose under the hood, and does not depend on the heroku CLI tools.

## Motivation and Context
Implements #2467 
This is a first step towards support for deploying to Heroku using docker containers.

## How Has This Been Tested?
The barlett1932 experiments runs. A new CI job creates the docker images and runs barlett1932 with bots.

## Docker hub space
To support this pull request we'll need to configure github actions to push to https://hub.docker.com/r/dallinger/dallinger
For now we're using https://hub.docker.com/r/dallingerimages/dallinger
The images were manually generated and uploaded. There is no mechanism to automate the upload yet, although most of it is in place already in Github CI to support tests.

## Prerequisites for testing
You need a recent (>= 18.09) version of `docker` installed on the system. 

The command
```
cd demos/dlgr/demos/bartlett193
dallinger docker debug
```
will start a docker environment. Use `--bot` option to use the `dallinger/dallinger-bot` docker image and run the bots using chromedriver inside the container.

I used the generated images to successfully do a manual deployment on docker based Heroku dynos, so this work is a good step towards #2556.

### CODE REVIEW

File: Dockerfile
Comment (alecpm) [inline Dockerfile:26]: I just added some chromedriver stuff to ci.yml in order to run bots now that PhantomJS is deprecated. I think it would probably makes sense to also include Chrome and Chromedriver in the docker build.

**Review by alecpm — CHANGES_REQUESTED** [review]
I like this refactoring, and now that I see it in code I think I agree with @jessesnyder that `dallinger docker debug` is probably a good way to handle these different types of deployments. `dallinger deploy-container` seems like a good idea to me too, but seems to perhaps be inconsistent with the debug command. It should be fine as long as we have a solid idea how these commands are extended, e.g. `dallinger deploy-container --sandbox --ecs`? Or maybe `dallinger ecs deploy` or `dallinger heroku-container deploy` in analogy with the `dallinger docker debug` command? This is probably a bit of premature bike-shedding, but I do think it's important we get this right. Also, I think this could probably use some tests for the new command and wrapper.

File: Dockerfile
Comment (silviot) [reply Dockerfile:26]: Including chrome and chromedriver would increase the size of the base docker image considerably. There are many reasons the image is best kept as lean as possible.

I need to look into bots to understand in which scenarios chrome and chromedriver are needed. But it's my understanding most of the time they're not.

If they're only needed during development we could maintain two docker images, `dallinger/dallinger` and `dallinger/dallinger-dev`.

Comment (silviot) [reply Dockerfile:26]: Adding chrome to the current image would increase image size by 412 Mb uncompressed. If you run this command to try and install it you'll see the long list of dependencies needed:
```
docker run --rm dallinger/dallinger bash -c '
apt update
apt install -y wget
wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb
SIZE_BEFORE=$(du -scx /)
echo Size before installing chrome: $SIZE_BEFORE
apt install -y --no-install-recommends ./google-chrome-stable_current_amd64.deb
echo Size before installing chrome: $SIZE_BEFORE
echo Size after installing chrome:
du -scx /
'
```
and the increase in the image size:
```
Size before installing chrome: 561840 / 561840 total
Size after installing chrome:
973840	/
973840	total
```

Comment (alecpm) [reply Dockerfile:26]: Generally speaking if an experiment uses bots (not many do, I think, but who knows) then the bots are run by a worker instance IIRC. Even when experiments do use bots, they don't necessarily require chromedriver.  There's a "high performance" bot base class that people can use which just makes API calls directly instead of going through the browser. So maybe it could be an option that's disabled by default? Probably nothing to worry about for now given the image size implications.

File: tests/conftest.py
Comment (alecpm) [inline tests/conftest.py:207]: nitpick: you probably didn't mean to leave this here

File: dallinger/docker/tools.py
Comment (alecpm) [inline dallinger/docker/tools.py:83]: This should probably look at the config variables `webdriver` and `recruiter` (there's a `"bots"` recruiter which is what `--bot` enables) to determine if it needs to install chrome or firefox, but I don't think that's critical at this stage. It gets more complicates with the `MultiRecruiter`, but I don't think we need to spend time on that right now.

File: dallinger/docker/tools.py
Comment (alecpm) [inline dallinger/docker/tools.py:161]: I don't think we've ever worried about the worker log stream for this purpose, so this should be fine.

File: docs/source/docker_setup.rst
Comment (alecpm) [inline docs/source/docker_setup.rst:9]: This definitely needs some more install and usage documentation.

Comment (alecpm) [reply docs/source/docker_setup.rst:9]: When I tried to run `dallinger docker debug` from `demos/dlgr/demos/bartlett1932` the `docker-compose` command failed during the `Dockerfile.worker` build with:
```
 => [2/4] COPY . /experiment                                               0.1s
 => [3/4] WORKDIR /experiment                                              0.0s
 => ERROR [4/4] RUN bash -c "python3 -m pip install -r <(grep -v ^dalling  0.4s
------
 > [4/4] RUN bash -c "python3 -m pip install -r <(grep -v ^dallinger requirements.txt)":
#8 0.375 /usr/bin/python3: No module named pip
------
executor failed running [/bin/sh -c bash -c "python3 -m pip install -r <(grep -v ^dallinger requirements.txt)"]: exit code: 1
ERROR: Service 'worker' failed to build
```
Not sure how that would have happened since `python3-pip` should be in the runtime image.

**Review by alecpm — CHANGES_REQUESTED** [review]
In general this looks good to me, especially the re-organization of the command line module. I couldn't get `dallinger docker debug` to work for me locally, which seemed to be some sort of problem with the `dallinger/dallinger:latest` image (see my comments for details). I think it should be OK to merge once it has a little more documentation and I can get it to run successfully.

File: dallinger/docker/tools.py
Comment (jessesnyder) [reply dallinger/docker/tools.py:83]: I think I disagree. I like the fact that the higher-level component (the `DockerDebugDeployment`) knows _why_ Chrome is needed and just tells the lower-level component (this one) to make it happen. The abstraction boundary leaks a little bit since the docker image is called "dallinger-bot", but nothing's perfect. :)

**Review by jessesnyder — COMMENTED** [review]
There's a ton progress here on providing the Docker functionality, and a lot of progress in re-working the broader deployment context to support it gracefully. The work that's been done makes the direction of further improvements more obvious, so it's tempting to keep doing more, but given the scope of change, I think we should do as little additional work as possible before merging, given that this feature doesn't affect existing functionality. We should just agree on what the minimum is: we can all run Bartlett locally after following documented instructions, maybe?

**Comment by jessesnyder** [comment]
## Test run report

- Installation ✅ 
- `dallinger docker debug` run [almost]
- shutdown ❓ 

Something funny happened with setting the hostname for the `/ad` route, with the browser opening "web:5000" (the dashboard opened correctly)

I don't find that vnc_auto.html file anywhere, and it raised a 404

Shutdown worked correctly, as far as I could tell, though there was a confusing message in the log ("WARNING: Network bartlett1932_default not found"):
```
❯❯ Experiment summary: {'status': 'success', 'summary': [['approved', 2]], 'completed': True, 'unfilled_networks': 0, 'nodes_remaining': 0, 'required_nodes': 0}

❯❯ Experiment completed, all nodes filled.
Stopping bartlett1932_worker_1     ... done
Stopping bartlett1932_web_1        ...
Stopping bartlett1932_postgresql_1 ... done
Stopping bartlett1932_web_1        ... done

❯❯ Completed debugging of experiment with id f53c7bc8-7840-adec-9f6d-19b32308febc
Removing bartlett1932_worker_1     ... done
Removing bartlett1932_web_1        ... done
Removing bartlett1932_postgresql_1 ... done
Removing bartlett1932_redis_1      ... done
Removing network bartlett1932_default
Removing network bartlett1932_default
WARNING: Network bartlett1932_default not found.
```

web_and_Dashboard_Home_-_Dashboard_and_Site_Error

File: dallinger/docker/docker-compose.yml.j2
Comment (silviot) [reply dallinger/docker/docker-compose.yml.j2]: I was missing `FLASK_SECRET_KEY` and needed to make it work.

Your comment made me realize that the old method of passing it to the local heroku process was via environment variable. Now I changed this to pass that var down to the docker container.

File: dallinger/deployment.py
Comment (silviot) [reply dallinger/deployment.py:583]: `HerokuLocalDeployment` currently calls the `__enter__` method of HerokuLocalWrapper` to start the services.

The services are started with a call to `subprocess.Popen` in the `_boot` method. That invocation relies on the CWD to be set correctly. What I meant with that comment was that that specific `os.chdir` has non obvious consequences  somewhere else far away. I should probably phrase it better.

**Comment by silviot** [comment]
@jessesnyder Thanks for the report!
I believe I broke the non-bot case when I made the bots work and added the test in the CI.
This makes me realise that the non-bot path needs some kind of tests too, but it will require some thought.

File: dallinger/docker/tools.py
Comment (alecpm) [reply dallinger/docker/tools.py:83]: You're right, but that I think that just means that at the `DockerDebugDeployment` we should be checking the `webdriver_type` and `recruiter` instead of relying on the `--bot` flag. For now it's probably fine as it is.

File: docs/source/docker_setup.rst
Comment (alecpm) [reply docs/source/docker_setup.rst:9]: Thanks! I had generated updated base images locally using `docker build`, but it's not clear to me how I would make the `dallinger docker debug` command use them. Maybe something to add to the documentation? There should probably be a clear way to build updated local base images and ensure that the `dallinger` command uses them.

File: dallinger/docker/tools.py
Comment (jessesnyder) [reply dallinger/docker/tools.py:83]: Ah... yep; agreed.

**Comment by alecpm** [comment]
Regarding a test for a successful docker launch without bots, one option would be to just run a bot in the test function after launching the docker command. You should be able to run a bot from a test using  e.g. the `bartlett_dir` fixture using something like:

```
from experiment import Bot

bot = Bot(url_to_docker_web_app, assignment_id, hit_id, worker_id)
bot.run_experiment()
```

However, I think simpler tests in `test_command_line` and `test_deployment` are more important than those sorts of full-stack integration tests. But more importantly, I agree with Jesse that this looks really great and provides a really good foundation. Apart from whatever needs to be done to make a debug run work without bots and a little more documentation, I think it's ready to merge.

File: docs/source/docker_setup.rst
Comment (silviot) [reply docs/source/docker_setup.rst:9]: @alecpm The images used are hardcoded. Now they are `dallingerimages/dallinger` and `dallingerimages/dallinger-bot`.
If you don't have them locally docker will try and download them from the public registry.

I'm very surprised you get that error. That's why my first thought was of you having the wrong image.
The image definitely ships `pip` for python 3.
Can you try to run `docker run --rm dallingerimages/dallinger python3 -m pip install nltk`?
For me it succeeds, but given your error above I expect it to fail on your machine.
Is this the sha of the image you're using?
```
docker inspect dallingerimages/dallinger|grep Id
        "Id": "sha256:4ef4b3db35e0d0a00495d95ced04fce16ba67757ba54944c690f283caabac5f0",
```

Comment (alecpm) [reply docs/source/docker_setup.rst:9]: @silviot This is probably all be down to me being a total Docker newbie. After running `docker build .` in the Dallinger dir (which does seem to build all the images), I end up with a single orphaned image:
```
% docker image ls 
REPOSITORY               TAG       IMAGE ID       CREATED          SIZE
<none>                   <none>    acb1300c29d5   39 seconds ago   2.5GB
```
Perhaps there's a build command I should be using to ensure that the build gets assigned to the correct repo?

**Comment by silviot** [comment]
@alecpm Thanks for the pointer! It helped a lot in writing a test where the test itself drives the bots.
I really wanted _something_ wrt testing that will break in case the functionality does not work anymore, as it happened when Jesse tested it and I just broke it ;-).

I agree more granular tests would be ideal, in a later iteration.

File: dallinger/docker/tools.py
Comment (silviot) [reply dallinger/docker/tools.py:88]: Got it, but leaving as is for now.

File: dallinger/docker/tools.py
Comment (silviot) [reply dallinger/docker/tools.py:168]: For me it was painful because I had errors in the worker and I needed to fish them out with (relatively) arcane `docker-compose` invocations. I believe we should talk a bit more about this.

File: docs/source/docker_setup.rst
Comment (silviot) [reply docs/source/docker_setup.rst]: @alecpm I included the needed commands in the docs now. Your build command was missing the tag to assign to the built image (passed with `-t`) and the target to build (our Dockerfile provides more than one target, and by default I believe docker will build the last one in the file, so the chromedriver image in this case) that you can specify with `--target`.

**Review by alecpm — APPROVED** [review]
Looks great! Thanks @silviot!

**Pull request merged**
